### DISCUSSION TITLE
Fix [ELSE] reference implementation to handle comment lines with conditional compilation words.

### DISCUSSION THREAD

**Discussion by SirWumpus** [ForthHub/discussion#143] [Ideas]
The Forth Draft 19.1 reference implementation for `[ELSE]` (section E.15.6.2.2531) should be improved to allow for single-line commented `[IF]`, `[ELSE]`, and `[THEN]` words.  I stumbled on this cause I had commented out some conditional compilation words while experimenting / tweaking some code and started to notice some odd behaviour.

Consider the following revised test cases:
```
\ Multiline with single line comment of [IF] [ELSE] [THEN] words, ie.
\ developer is futzing around with code and temorarily comments words.
T{ TRUE [IF]
    \ TRUE [IF]
        1
    \ [ELSE]
        2
    [ELSE]
        3
    \ [THEN]
        4
    [THEN] -> 1 2 }T

T{ FALSE [IF]
        1
    \ [ELSE]
        2
    [ELSE]
    \ TRUE [IF}
        3
    \ [THEN]
        4
    [THEN] -> 3 4 }T
```

The following revised `[ELSE]` solves the above:
```
: [ELSE] ( -- )
	1 BEGIN 				\ level
	  BEGIN PARSE-NAME DUP WHILE		\ level adr len
	    2DUP S" \" COMPARE 0= IF
	      \ Ignore remainder of comment line.
	      2DROP POSTPONE \
	    ELSE 2DUP S" [IF]" COMPARE 0= IF		\ level adr len
	      2DROP 1+				\ level'
	    ELSE				\ level adr len
	      2DUP S" [ELSE]" COMPARE 0= IF	\ level adr len
	        2DROP 1-			\ level'
	        \ Not yet zero, then restore previous level while nested.
	        DUP IF 1+ THEN			\ level'
	      ELSE 				\ level adr len
	        S" [THEN]" COMPARE 0= IF	\ level
	          1-				\ level'
	        THEN
	      THEN
	    THEN THEN
	    ?DUP 0= IF EXIT THEN		\ level'
	  REPEAT 2DROP				\ level
	REFILL 0= UNTIL				\ level
	DROP
; IMMEDIATE
```

*Note the above does not handle parenthesis comments (which I tend not to use if possible).*

**Comment by ruv** [comment]
This problem cannot be solved in Forth.

A user can redefine line comment (or introduce an own variant):
```
: \G   \ special handling ... 
  postpone \
; immediate
```

Or `[else]`  can be used in a string literal like `s" .... [else] ... "` inside a containing `[if] [then]`.

It's confused if these words can be commented by `\`, but things become broken if they are appeared in other comments or string literals.

So they should be considered as a top-level markup, which should not appear in comments or string literals.

This problem was also [discussed on forth-standard.org](https://forth-standard.org/standard/tools/BracketELSE#contribution-74).

**Reply by SirWumpus** [reply]
Ah.  Well the only reason I had to comment these words out was for temporary experimentation and testing  (which is how I stumbled on the issue) and then either uncomment or remove them once a course of action was decided on.   I'm just used to how in C I can comment out `#ifdef` ... `#endif` lines and insert temporary replacements while toying with the code.  Its not something I'd leave in a final working solution.  

As with all things Forth, if you redefine standard / common words, you've probably gone off the reservation on some tangent and to expect there be dragons.

Reading [that discussion](https://forth-standard.org/standard/tools/BracketELSE#contribution-74) goes to show I'm not the first, but could have been the last if this had been "clarified" in the rationale so as to avoid the mine field.
